server/v2/cometbft/abci.go

```diff
@@ -446,6 +486,29 @@ func (c *Consensus[T]) ProcessProposal(
 func (c *Consensus[T]) FinalizeBlock(
 	ctx context.Context,
 	req *abciproto.FinalizeBlockRequest,
+) (*abciproto.FinalizeBlockResponse, error) {
+	if c.optimisticExec.Initialized() {
+		// check if the hash we got is the same as the one we are executing
+		aborted := c.optimisticExec.AbortIfNeeded(req.Hash)
+
+		// Wait for the OE to finish, regardless of whether it was aborted or not
+		res, err := c.optimisticExec.WaitResult()
+
+		// only return if we are not aborting
+		if !aborted {
+			return res, err
+		}
+
+		// if it was aborted, we need to reset the state
+		c.optimisticExec.Reset()
+	}
+
+	return c.internalFinalizeBlock(ctx, req)
```

Comment:
Can't comment on the right line below, but `internalFinalizeBlock` should not commit to store. I think internalFinalizeBlock should return stateChanges, then everything else should be done on the actual FinalizeBlock call.
This is because there could be an equivocation and we could be running the wrong finalize block request